File: library/src/main/java/com/owncloud/android/lib/resources/albums/CreateNewAlbumRemoteOperation.java
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2025 TSI-mc <surinder.kumar@t-systems.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package com.owncloud.android.lib.resources.albums;

import com.nextcloud.common.SessionTimeOut;
import com.nextcloud.common.SessionTimeOutKt;
import com.owncloud.android.lib.common.OwnCloudClient;
import com.owncloud.android.lib.common.network.WebdavUtils;
import com.owncloud.android.lib.common.operations.RemoteOperation;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.lib.common.utils.Log_OC;

import org.apache.jackrabbit.webdav.client.methods.MkColMethod;

public class CreateNewAlbumRemoteOperation extends RemoteOperation<Void> {
    private static final String TAG = CreateNewAlbumRemoteOperation.class.getSimpleName();
    private final String newAlbumName;
    private final SessionTimeOut sessionTimeOut;

    public CreateNewAlbumRemoteOperation(String newAlbumName) {
        this(newAlbumName, SessionTimeOutKt.getDefaultSessionTimeOut());
    }

    public CreateNewAlbumRemoteOperation(String newAlbumName, SessionTimeOut sessionTimeOut) {
        this.newAlbumName = newAlbumName;
        this.sessionTimeOut = sessionTimeOut;
    }

    public String getNewAlbumName() {
        return newAlbumName;
    }

    /**
     * Performs the operation.
     *
     * @param client Client object to communicate with the remote ownCloud server.
     */
    @Override
    protected RemoteOperationResult<Void> run(OwnCloudClient client) {
        MkColMethod mkCol = null;
        RemoteOperationResult<Void> result;
        try {
            mkCol = new MkColMethod(client.getBaseUri() + "/remote.php/dav/photos/" + client.getUserId() + "/albums" + WebdavUtils.encodePath(newAlbumName));
            client.executeMethod(mkCol, sessionTimeOut.getReadTimeOut(), sessionTimeOut.getConnectionTimeOut());
            if (405 == mkCol.getStatusCode()) {
                result = new RemoteOperationResult<>(RemoteOperationResult.ResultCode.FOLDER_ALREADY_EXISTS);
            } else {
                result = new RemoteOperationResult<>(mkCol.succeeded(), mkCol);
                result.setResultData(null);
            }

            Log_OC.d(TAG, "Create album " + newAlbumName + ": " + result.getLogMessage());
            client.exhaustResponse(mkCol.getResponseBodyAsStream());
        } catch (Exception e) {
            result = new RemoteOperationResult<>(e);
            Log_OC.e(TAG, "Create album " + newAlbumName + ": " + result.getLogMessage(), e);
        } finally {
            if (mkCol != null) {
                mkCol.releaseConnection();
            }

        }

        return result;
    }

}

File: library/src/main/java/com/owncloud/android/lib/resources/albums/CreateNewAlbumRemoteOperation.kt
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2025 TSI-mc <surinder.kumar@t-systems.com>
 * SPDX-License-Identifier: MIT
 */
package com.owncloud.android.lib.resources.albums

import com.nextcloud.common.SessionTimeOut
import com.nextcloud.common.defaultSessionTimeOut
import com.owncloud.android.lib.common.OwnCloudClient
import com.owncloud.android.lib.common.operations.RemoteOperation
import com.owncloud.android.lib.common.operations.RemoteOperationResult
import com.owncloud.android.lib.common.operations.RemoteOperationResult.ResultCode
import com.owncloud.android.lib.common.utils.Log_OC
import org.apache.commons.httpclient.HttpStatus
import org.apache.jackrabbit.webdav.client.methods.MkColMethod

class CreateNewAlbumRemoteOperation
    @JvmOverloads
    constructor(
        val newAlbumName: String,
        private val sessionTimeOut: SessionTimeOut = defaultSessionTimeOut
    ) : RemoteOperation<Void>() {
        @Deprecated("Deprecated in Java")
        @Suppress("TooGenericExceptionCaught", "DEPRECATION")
        override fun run(client: OwnCloudClient): RemoteOperationResult<Void> {
            var mkCol: MkColMethod? = null
            return try {
                mkCol = MkColMethod(client.albumUri(newAlbumName))
                client.executeMethod(mkCol, sessionTimeOut.readTimeOut, sessionTimeOut.connectionTimeOut)

                val result =
                    if (mkCol.statusCode == HttpStatus.SC_METHOD_NOT_ALLOWED) {
                        RemoteOperationResult(ResultCode.FOLDER_ALREADY_EXISTS)
                    } else {
                        RemoteOperationResult<Void>(mkCol.succeeded(), mkCol)
                    }
                Log_OC.d(TAG, "Create album $newAlbumName : ${result.logMessage}")
                client.exhaustResponse(mkCol.responseBodyAsStream)
                result
            } catch (e: Exception) {
                failure(e)
            } finally {
                mkCol?.releaseConnection()
            }
        }

        @Suppress("DEPRECATION")
        private fun failure(e: Exception): RemoteOperationResult<Void> =
            RemoteOperationResult<Void>(e).also {
                Log_OC.e(TAG, "Create album $newAlbumName : ${it.logMessage}", e)
            }

        companion object {
            private val TAG: String = CreateNewAlbumRemoteOperation::class.java.simpleName
        }
    }
